lib/date-string.js

```diff
@@ -0,0 +1,39 @@
+// Copyright IBM Corp. 2013,2016. All Rights Reserved.
+// Node module: loopback-datasource-juggler
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+'use strict';
+
+module.exports = DateString;
+
+/**
+ * A String whose value is a valid representation of a Date.
+ * Use this type if you need to preserve the format of the value and still
+ * check if it's valid.
+ * @param {String} value
+ * @constructor
+ */
+function DateString(value) {
+  if (!(this instanceof DateString)) {
+    return new DateString(value);
+  }
+
+  if (typeof(value) !== 'string') {
+    throw new Error('Input must be a string');
+  }
+  // Preserve the reference for return via .toDate
+  this.date = new Date(value);
+  if (isNaN(this.date.getTime())) {
+    throw new Error('Invalid date');
+  }
+  this.value = value;
+};
+
+/**
+ * Return the equivalent Date object for this DateString.
+ * @returns {Date} A JavaScript Date object
+ */
+DateString.prototype.toDate = function() {
```

Comment:
IMO, this function needs to be parameterized by the timezone we want to get the value in. Otherwise you can end up with the same problem when somebody calls `toLocaleString` as you are trying to fix: _if you were attempting to insert a birthday into a database, you wouldn't want the timezone offset to push that day forwards or backwards, right?_

```
$ node
> new Date('2017-04-27').toLocaleString(undefined, {timeZone:'America/Toronto'})
'4/26/2017, 8:00:00 PM'
```

The snippet above shows what happens to value `2017-04-27` when it's displayed via `toLocaleString()` by a computer located in Toronto (think of  `new DateString('2017-04-27').toDate().toLocaleString()`). Because my timezone is `Europe/Prague`, I had to explicitly tell `toLocaleString` to use a different one.

I think it may be better to not convert `DateTime` values to `Date` at all and let each connector to decide how to represent a date value.